### ISSUE TITLE
ML Project

### ISSUE THREAD

**Issue by Mahyarbo** [Mahyarbo/Tennis-Talent-Detection-Project-#1]
User (App / Video)
        |
Frontend (Simple UI / Streamlit)
        |
ML Backend (Python)
 ├── Text-based ML Model
 ├── Pose Detection (MediaPipe)
 ├── Feature Extraction
 ├── Talent Scoring Model

tennis_talent_ai/
│
├── app.py                 # UI اپ
├── pose_analysis.py       # Pose Detection
├── feature_extraction.py  # استخراج ویژگی
├── model.py               # ML Model
├── train.py               # آموزش مدل
├── dataset.csv            # داده آموزشی
├── requirements.txt

numpy
pandas
scikit-learn
mediapipe
opencv-python
streamlit

pip install -r requirements.txt

age,height,weight,flexibility,speed,strength,elbow_angle,knee_angle,label
18,175,70,8,7,6,155,130,1
22,182,78,6,6,7,145,120,1
19,168,65,9,8,5,160,140,1
25,170,85,4,4,8,120,100,0
30,165,90,3,3,6,110,95,0

import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
import joblib

data = pd.read_csv("dataset.csv")

X = data.drop("label", axis=1)
y = data["label"]

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

model = RandomForestClassifier(
    n_estimators=200,
    max_depth=10,
    random_state=42
)

model.fit(X_train, y_train)

joblib.dump(model, "tennis_model.pkl")

print("Model trained and saved")

import cv2
import mediapipe as mp
import numpy as np

mp_pose = mp.solutions.pose
pose = mp_pose.Pose()

def calculate_angle(a, b, c):
    a, b, c = np.array(a), np.array(b), np.array(c)
    ba = a - b
    bc = c - b
    angle = np.degrees(
        np.arccos(
            np.dot(ba, bc) / (np.linalg.norm(ba) * np.linalg.norm(bc))
        )
    )
    return angle

def analyze_video(video_path):
    cap = cv2.VideoCapture(video_path)

    elbow_angles = []
    knee_angles = []

    while cap.isOpened():
        ret, frame = cap.read()
        if not ret:
            break

        rgb = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
        result = pose.process(rgb)

        if result.pose_landmarks:
            lm = result.pose_landmarks.landmark

            shoulder = [lm[12].x, lm[12].y]
            elbow = [lm[14].x, lm[14].y]
            wrist = [lm[16].x, lm[16].y]

            hip = [lm[24].x, lm[24].y]
            knee = [lm[26].x, lm[26].y]
            ankle = [lm[28].x, lm[28].y]

            elbow_angles.append(calculate_angle(shoulder, elbow, wrist))
            knee_angles.append(calculate_angle(hip, knee, ankle))

    cap.release()

    return np.mean(elbow_angles), np.mean(knee_angles)

def build_feature_vector(
    age, height, weight, flexibility, speed, strength,
    elbow_angle, knee_angle
):
    return [[
        age, height, weight,
        flexibility, speed, strength,
        elbow_angle, knee_angle
    ]]

import joblib

model = joblib.load("tennis_model.pkl")

def predict_talent(features):
    prediction = model.predict(features)[0]
    probability = model.predict_proba(features)[0][1]

    return prediction, probability

import streamlit as st
from pose_analysis import analyze_video
from feature_extraction import build_feature_vector
from model import predict_talent

st.set_page_config(page_title="Tennis Talent AI", layout="centered")

st.title("Tennis Talent Identification App")

age = st.slider("Age", 10, 40, 20)
height = st.slider("Height (cm)", 140, 200, 175)
weight = st.slider("Weight (kg)", 40, 120, 70)

flexibility = st.slider("Flexibility (1–10)", 1, 10, 5)
speed = st.slider("Speed (1–10)", 1, 10, 5)
strength = st.slider("Strength (1–10)", 1, 10, 5)

video = st.file_uploader("Upload Tennis Video", type=["mp4"])

if st.button("Analyze Talent"):
    if video:
        with open("temp.mp4", "wb") as f:
            f.write(video.read())

        elbow_angle, knee_angle = analyze_video("temp.mp4")

        features = build_feature_vector(
            age, height, weight,
            flexibility, speed, strength,
            elbow_angle, knee_angle
        )

        pred, prob = predict_talent(features)

        if pred == 1:
            st.success(f"High Tennis Talent ({prob*100:.1f}%)")
        else:
            st.warning(f"Low Tennis Suitability ({prob*100:.1f}%)")

        st.write("Elbow Angle:", elbow_angle)
        st.write("Knee Angle:", knee_angle)

streamlit run app.py
